lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -2093,6 +2093,14 @@ bool TypeChecker::diagnoseExplicitUnavailability(
     return false;
   }
 
+  // Calling unavailable code from within code with the same
+  // unavailability is OK -- the eventual caller can't call the
+  // enclosing code in the same situations it wouldn't be able to
+  // call this code.
+  if (isInsideUnavailableDeclaration(R, DC)) {
```

Comment:
I think this is still a problem: it would let something guarded with `@available(iOSApplicationExtension, unavailable)` call things that were `@available(iOS, unavailable)`. Modeling application extension restrictions as availability really needs more attention in Swift.